src/lfortran/semantics/ast_common_visitor.h

```
@@ -2929,7 +2929,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     }
                     init_expr = ASRUtils::EXPR(tmp);
                     value = ASRUtils::expr_value(init_expr);
-
+                    if(value->type==63 && type->type==0){
```

Please use the `ASRUtils::expr_type()` utility function to get the type of an expression. 

I think we need to throw this error after comparing the number of dimensions (rank) of the LHS and RHS. You can use the `ASRUtils::extract_n_dims_from_ttype()` utility function to do this. 

Let's do this:
1. Find the place where we assign the values.
2. Compare the number of dimension of the LHS and RHS.
3. Throw error if not equal.

Do the same for declaration also, inside `visit_DeclarationUtil()`.